### PR TITLE
Parsing of file tags: Fix some subtle bugs and inconsistencies when reloading tags

### PR DESCRIPTION
**Pull request by uklotzde**
~~_Some tests are needed and I have to think about how to correctly handle cover art_~~

I'm constantly using some experimental functionality contained in the following 2 branches for updating and reloading file tags:

Write track metadata back into file tags:
https://github.com/uklotzde/mixxx/tree/WriteAudioTags

Implicitly reload and update track metadata from file tags:
https://github.com/uklotzde/mixxx/tree/ReloadAudioTags
(based on the previous _WriteAudioTags_ branch)

The latter branch treats file tags as the _book of records_ while the Mixxx library only serves as a cache for searching/filtering/organizing with additional track metadata like waveforms and cue points. Especially with this branch I was experiencing some unexpected behaviour.

Modified file tags are written with a delay when the in-memory track object goes out of scope. By loading or editing a modified track object again before the file tags could be updated in-memory metadata was lost and replaced with the latest version stored in the file. This has been fixed now by respecting the dirty flag of track objects appropriately. The dirty flag is only ignored when the user explicitly instructs Mixxx to reload the metadata from files.

*Changes*
* Tags are now parsed implicitly in the constructor of SoundSourceProxy, i.e. when accessing the corresponding file of a track object. No separate invocation of any function required, just specify the optional parameter.
* Reduced the number of functions for parsing file tags to a single function
* Now only a single bitmask parameter is needed to specify which and how tags are parsed and handled
* Logging has been improved for transparency and traceability

### CODE REVIEW

File: src/widget/wtracktableview.cpp [hunk 1]
```diff
@@ -1389,7 +1389,12 @@ void WTrackTableView::slotReloadTrackMetadata() {
     for (const QModelIndex& index : indices) {
         TrackPointer pTrack = trackModel->getTrack(index);
         if (pTrack) {
-            SoundSourceProxy(pTrack).loadTrackMetadata(true);
+            // The user has explicitly requested to reload metadata from the file
+            // to override the information within Mixxx!
+            SoundSourceProxy(pTrack,
+                    SoundSourceProxy::PARSE_METADATA |
+                    SoundSourceProxy::RELOAD_METADATA_EVEN_IF_ALREADY_PARSED |
+                    SoundSourceProxy::RELOAD_METADATA_EVEN_IF_DIRTY);
```
Comment (daschuer) [inline src/widget/wtracktableview.cpp]: Both later flags are always set as pair, is there a future use case for only one of them? 
If not, we can use a single "force" flag or PARSE_ENSURE_METADATA / PARSE_RELOAD_METADATA

File: src/sources/soundsourceproxy.cpp [hunk 2]
```diff
@@ -428,75 +435,90 @@ namespace {
     }
 } // anonymous namespace
 
-void SoundSourceProxy::loadTrackMetadataAndCoverArt(
-        bool withCoverArt,
-        bool reloadFromFile) const {
+void SoundSourceProxy::parseFileTags(int parseOptions) {
     DEBUG_ASSERT(m_pTrack);
 
+    if (!(parseOptions & PARSE_METADATA_AND_COVERART)) {
```
Comment (daschuer) [inline src/sources/soundsourceproxy.cpp]: I think it is easier to read if you use the single flags here. I was a short moment confused about the "AND"

**Review by daschuer — APPROVED** [review]
Thank you, It LGTM. 
Only some questions.

File: src/widget/wtracktableview.cpp (hunk 1 above)
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: The later use case is in the _ReloadAudioTags_ branch where reloading is only skipped if the track is dirty. In this case it doesn't matter if the tags have already been parsed once.

**Comment by uklotzde** [comment]
TODO
* I noticed that we can get rid of a lot of temporary track objects by using SoundSourceProxy::parseTrackMetadata().
* Cover art is never reloaded by the SoundSourceProxy constructor once it has been parsed. Reloading of cover art is done separately.
* Both metadata and cover art must loaded together, otherwise cover art might be ignored entirely (see previous bullet point)

File: src/sources/soundsourceproxy.cpp (hunk 2 above)
Comment (uklotzde) [inline src/sources/soundsourceproxy.cpp]: The single flags will become private, see my TODO comment ;)

**Comment by daschuer** [comment]
Ok, should I mereg this now?

**Comment by uklotzde** [comment]
I just discovered that everything breaks down into 3 options for the constructor of SoundSourceProxy:
* IGNORE_FILE_TAGS
* PARSE_FILE_TAGS_ONCE
* PARSE_FILE_TAGS_AGAIN

**Comment by uklotzde** [comment]
No, please wait with merging until I remove the [WiP] tag! The new options just became obvious while trying to add some simple test cases ;)

**Comment by uklotzde** [comment]
@daschuer I need to work on the whole design. Setting properties by a side effect in SoundSourceProxy is at least unintuitive.

**Comment by uklotzde** [comment]
@daschuer: I now consider this finished and tested. Even cover art can be reloaded under certain conditions that ensure not to overwrite custom selections made by users.

**Comment by daschuer** [comment]
Nice! Thank you very much! LGTM.

**Pull request merged**